crates/re_space_view_spatial/src/space_view_3d.rs

```diff
@@ -71,46 +74,56 @@ impl SpaceViewClass for SpatialSpaceView3D {
-        // We already know all the entities that have a pinhole camera - indirectly today through visualizers, but could also be directly.
-        // Meaning we don't need to walk until we find a pinhole camera!
-        // Obviously should skip the whole thing if there are no pinhole cameras under space_origin!
-        // TODO(jleibs): Component prefix tree on EntityTree would fix this problem nicely.
-
-        let mut entities_under_pinhole = IntSet::default();
-
-        fn visit_children_recursively(
-            tree: &EntityTree,
-            entities_under_pinhole: &mut IntSet<EntityPathHash>,
-        ) {
-            if tree
-                .entity
-                .components
-                .contains_key(&PinholeProjection::name())
-            {
-                // This and all children under it are under a pinhole camera!
-                tree.visit_children_recursively(&mut |ent_path| {
-                    entities_under_pinhole.insert(ent_path.hash());
-                });
-            } else {
-                for child in tree.children.values() {
-                    visit_children_recursively(child, entities_under_pinhole);
+        // TODO(andreas): The `VisualizableFilterContext` depends entirely on the spatial topology.
+        // If the topology hasn't changed, we don't need to recompute any of this.
+        // Also, we arrive at the same `VisualizableFilterContext` for lots of different origins!
+
+        let context = SpatialTopology::access(entity_db.store_id(), |topo| {
+            let primary_space = topo.subspace_for_entity(space_origin);
+            match primary_space.dimensionality {
+                SubSpaceDimensionality::Unknown => VisualizableFilterContext3D {
+                    entities_in_main_3d_space: primary_space.entities.clone(),
+                    entities_under_pinholes: Default::default(),
+                },
+
+                SubSpaceDimensionality::ThreeD => {
+                    // All entities in the 3d space are visualizable + everything under pinholes.
+                    let mut entities_in_main_3d_space = primary_space.entities.clone();
+                    let mut entities_under_pinholes = IntSet::<EntityPath>::default();
+
+                    for (child_origin, connection) in &primary_space.child_spaces {
+                        if connection.is_connected_pinhole() {
+                            let Some(child_space) =
+                                topo.subspace_for_subspace_origin(child_origin.hash())
+                            else {
+                                // Should never happen, implies that a child space is not in the list of subspaces.
+                                continue;
+                            };
+
+                            // Entities _at_ pinholes are a special case: we display both 3d and 2d visualizers for them.
+                            entities_in_main_3d_space.insert(child_space.origin.clone());
```

Comment:
Let's not change it yet but this continues to make me think these entities should just be in `primary_space.entities` in the first place so we don't have to mutate the set.

Reply:
yeah we can totally throw the responsibility for this onto the space topology, just need to add a few more unit tests that it does the right thing under updates.
Nice thing is that unlike the old SpaceInfo we know what we're dealing with and who's the "target audience"